funderfinder/utils/list_gsoc.py

```diff
@@ -0,0 +1,263 @@
+    :return: A generator of dicts containing project metadata
+    """
+    listing_page = requests.get(link)
+    soup = bs4.BeautifulSoup(listing_page.text, features="html.parser")
+    project_containers = get_early_archive_listing_links(soup)
+    year = int(link.strip().strip("/").split("/")[-1])
+    for container in project_containers:
+        meta = get_early_archive_project(container, year)
+        yield meta
+        time.sleep(SCRAPE_DELAY)
+
+
+def get_projects_before_2016() -> iter:
+    """
+    Retrieves project metadata from before 2016. GSOC uses an older website format for these
+    :return: A generator of dicts containing project metadata
+    """
+    listing_page = requests.get("https://www.google-melange.com/archive/gsoc")
+    soup = bs4.BeautifulSoup(listing_page.text, features="html.parser")
+    year_link_containers = get_early_archive_listing_links(soup)
+    for container in year_link_containers:
+        year_link = extract_listing_link(container)
+        if not year_link:
+            continue
+        print(f"Getting projects for {year_link}")
+        year_projects = get_early_archive_year_projects(year_link)
+        for project in year_projects:
+            yield project
+
+
+def get_modern_archive_project(year: int, slug: str) -> dict:
+    """
+    Retrieves project metadata for a project and year after 2016.
+    :param year: Year we want to retreive project metadata from
+    :param slug: The GSOC project slug, retrieved from their API
+    :return: Dict of project metadata
+    """
+    time.sleep(SCRAPE_DELAY)
+    project_url = f"https://summerofcode.withgoogle.com/api/archive/programs/{year}/organizations/{slug}/"
+    meta = requests.get(project_url).json()
+    repos = []
+    repos.extend(get_link_matches(meta["description_html"]))
+    repos.extend(get_link_matches(meta["ideas_list_url"]))
+    for student_project in meta["projects"]:
+        repos.extend(get_link_matches(student_project["abstract_html"]))
+        repos.extend(get_link_matches(student_project["project_code_url"]))
+    return {
+        "name": meta["name"],
+        "link": project_url,
+        "repos": list(set(repos)),
+        "year": year,
+    }
+
+
+def get_modern_archive_projects(year: int) -> tuple:
+    """
+    Retrieves project metadata for a year after 2016, if available
+    :param year: Year to retrieve project metadata from
+    :return: A tuple, with the first element a boolean which is True if we found project metadata for the year and
+    false otherwise (which it may be for the current year), and the second element an iterable of project metadata
+    """
+    org_url = f"https://summerofcode.withgoogle.com/api/archive/programs/{year}/organizations/"
+    orgs = requests.get(org_url).json()
+    is_ok = (type(orgs) == list) and (len(orgs) > 0)
+    if not is_ok:
+        return False, []
+    meta = (get_modern_archive_project(year, org["slug"]) for org in orgs)
+    return True, meta
```

Comment:
My Python instincts are fighting against this tuple return structure. While everything is correct here, I don't think I understand the need. Why can't a consumer of this function simply check that the returned value is falsy, i.e. empty? You probably have a good reason. If this was Go, I wouldn't think twice, but I don't see this construct often in Python.